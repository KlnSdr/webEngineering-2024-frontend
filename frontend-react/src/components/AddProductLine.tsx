import React, { useState } from "react";
import Dropdown from "./Dropdown";
import Button from "react-bootstrap/Button";
import Stack from "react-bootstrap/Stack";

interface AddProductLineProps {
  products: string[];
  initialValue: string;
  onChange: (product: string, amount: number) => void;
  onRemove: () => void;
  getUnitOf: (product: string) => string;
}

const AddProductLine: React.FC<AddProductLineProps> = ({
  products,
  initialValue,
  onChange,
  onRemove,
  getUnitOf,
}) => {
  const [unit, setUnit] = useState(getUnitOf(initialValue));
  const [amount, setAmount] = useState(0);
  const [product, setProduct] = useState(initialValue);

  return (
    <Stack direction="horizontal">
      <Dropdown
        options={products}
        initialValue={product}
        onChange={(val: string) => {
          setProduct(val);
          setUnit(getUnitOf(val));
          onChange(val, amount);
        }}
      />
      <input
        type="number"
        value={amount}
        onChange={(e) => {
          setAmount(parseInt(e.target.value));
          onChange(product, parseInt(e.target.value));
        }}
        className="form-control"
      />
      <label>{unit || getUnitOf(product)}</label>
      <Button
        onClick={() => {
          onRemove();
        }}
        className="bi bi-x"
      ></Button>
    </Stack>
  );
};

export default AddProductLine;
