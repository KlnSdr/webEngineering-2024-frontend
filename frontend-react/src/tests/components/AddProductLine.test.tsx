import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import '@testing-library/jest-dom/extend-expect';
import AddProductLine from "../../components/AddProductLine";

jest.mock("../../components/Dropdown", () => ({
                                   options,
                                   initialValue,
                                   onChange
                               }: any) => (
    <select
        data-testid="dropdown"
        defaultValue={initialValue}
        onChange={(e) => onChange(e.target.value)}
    >
        {options.map((option: string) => (
            <option key={option} value={option}>{option}</option>
        ))}
    </select>
));

describe("AddProductLine Component", () => {
    const products = ["Product 1", "Product 2", "Product 3"];
    const initialValue = "Product 1";
    const getUnitOf = (product: string) => (product === "Product 1" ? "kg" : "ltr");
    const getProduct = () => ({id: 1, productName: "Product 1", amount: 0 });

    test("renders with initial values", () => {
        const handleChange = jest.fn();
        const handleRemove = jest.fn();

        const {asFragment} = render(
            <AddProductLine
                products={products}
                initialValue={initialValue}
                onChange={handleChange}
                onRemove={handleRemove}
                getUnitOf={getUnitOf}
                getProduct={getProduct}/>
        );

        expect(screen.getByTestId("dropdown")).toHaveValue(initialValue);
        expect(screen.getByRole("spinbutton")).toHaveValue(0);
        expect(asFragment()).toMatchSnapshot();
    });

    test("updates product selection", () => {
        const handleChange = jest.fn();
        const handleRemove = jest.fn();

        render(
            <AddProductLine
                products={products}
                initialValue={initialValue}
                onChange={handleChange}
                onRemove={handleRemove}
                getUnitOf={getUnitOf}
                getProduct={getProduct}
            />
        );

        const dropdown = screen.getByTestId("dropdown");
        fireEvent.change(dropdown, { target: { value: "Product 2" } });

        expect(dropdown).toHaveValue("Product 2");
        expect(handleChange).toHaveBeenCalledWith("Product 2", 0);
        expect(screen.getByText("ltr")).toBeInTheDocument();
    });

    test("updates amount", () => {
        const handleChange = jest.fn();
        const handleRemove = jest.fn();

        render(
            <AddProductLine
                products={products}
                initialValue={initialValue}
                onChange={handleChange}
                onRemove={handleRemove}
                getUnitOf={getUnitOf}
                getProduct={getProduct}
            />
        );

        const amountInput = screen.getByRole("spinbutton");
        fireEvent.change(amountInput, { target: { value: "5" } });

        expect(amountInput).toHaveValue(5);
        expect(handleChange).toHaveBeenCalledWith("Product 1", 5);
    });

    test("calls onRemove callback", () => {
        const handleChange = jest.fn();
        const handleRemove = jest.fn();

        render(
            <AddProductLine
                products={products}
                initialValue={initialValue}
                onChange={handleChange}
                onRemove={handleRemove}
                getUnitOf={getUnitOf}
                getProduct={getProduct}
            />
        );

        const removeButton = screen.getByRole("button");
        fireEvent.click(removeButton);

        expect(handleRemove).toHaveBeenCalledTimes(1);
    });
});
