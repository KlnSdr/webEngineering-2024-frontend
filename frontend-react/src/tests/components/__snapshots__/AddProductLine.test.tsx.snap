// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddProductLine Component renders with initial values 1`] = `
<DocumentFragment>
  <div
    class="hstack"
  >
    <select
      data-testid="dropdown"
    >
      <option
        selected=""
        value="Product 1"
      >
        Product 1
      </option>
      <option
        value="Product 2"
      >
        Product 2
      </option>
      <option
        value="Product 3"
      >
        Product 3
      </option>
    </select>
    <input
      class="form-control"
      type="number"
      value="0"
    />
    <label>
      kg
    </label>
    <button
      class="bi bi-x btn btn-primary"
      type="button"
    />
  </div>
</DocumentFragment>
`;
